### PR TITLE
Construct `PolygonArray` or `MultiPolygonArray` from Exterior Coordinates 

### LINKED ISSUE #122: Faster construction of `PolygonArray` and `MultiPolygonArray`
I am working with a relatively large unstructured grid (57,000,000 polygons) that I am aiming to visualize. I am storing my geometries in a `spatialpandas.GeoDataFrame` with a single `MultiPolygonArray` column. However, the time that it takes me to construct these lead to about a 6-7x slowdown over using a `geopandas.GeoDataFrame` with the same data stored.

#### Code / Algorithm 
```Python
import uxarray as ux
import numpy as np
import spatialpandas
import geopandas 
import antimeridian
from shapely import polygons as Polygons

def to_gdf(uxgrid):

    # construct an array of shapely polygons
    polygons = Polygons(uxgrid.polygon_shells)
    
    # indices of polygons that need to be corrected for the antimeridian case
    antimeridian_indices = np.argwhere(np.any(np.abs(np.diff(uxgrid.polygon_shells[:, :, 0])) >= 180, axis=1)).squeeze()
    
    # obtain antimeridian polygons
    antimeridian_polygons = polygons[antimeridian_indices]
    
    # correct antimeridian polygons, stored as MultiPolygons
    corrected_polygons = [antimeridian.fix_polygon(P) for P in antimeridian_polygons]
    
    # replace original polygon with the corrected version
    for i in reversed(antimeridian_indices):
        polygons[i] = corrected_polygons.pop()
    
    ## ---- spatialpandas
    geometry = MultiPolygonArray(polygons)
    gdf = spatialpandas.GeoDataFrame({"geometry": geometry})  

    ## ---- geopandas
    # gdf = geopandas.GeoDataFrame({"geometry": polygons})  
    
    return gdf
```

#### Timings
| Method / # of Polygons | 1 ,791 | 7,153 | 28,57 | 236,853 | 56,623,106 | 
| ------------- | ------------- | ------------- | ------------- | ------------- | ------------- | 
| SpatialPandas  | 0.117s | 0.401s | 1.61s | 12.4s | 2851s | 
| GeoPandas  |  0.043s | 0.111s | 0.553s | 2.53s | 434s | 

Taken from a single NCAR Casper Node with 256gb of memory (dataset was able to fit into memory each time, no dask used)

#### Question
My key question is whether there's a suggested or faster way of constructing `PolygonArray` and `MultiPolygonArray` or better ways of handling such large datasets. Thanks!

### PR DESCRIPTION
**Pull request by philipc2**
Closes #122 

## Overview
- Introduces the `PolygonArray.from_exterior_coords()` and `MultiPolygonArray.from_exterior_coords()` class methods, which take-in an list/array of exterior coordinates equivalent to the following which are derived from Shapely Polygons/MultiPolygons

https://github.com/holoviz/spatialpandas/blob/dd941f2f5f6994c6f09996e9b8af315e16d00051/spatialpandas/geometry/polygon.py#L40-L42

https://github.com/holoviz/spatialpandas/blob/dd941f2f5f6994c6f09996e9b8af315e16d00051/spatialpandas/geometry/multipolygon.py#L40-L48

## Usage

In [UXarray](https://github.com/UXARRAY/uxarray), we use SpatialPandas paired with Datashader to visualize geoscience data that resides on unstructured grids as Polygons. Our unstructured grid is represented as a set of connectivity and coordinate arrays, which we would like to avoid needing to convert to Shapely, because we can derive the exterior of each Polygon quickly using the connectivity and coordinate information. 

## Performance Comparison
_Initial benchmarks, comparing across `n_polygon` & `n_vertices`_

All benchmarks are run on an [NCAR Casper HPC Node](https://arc.ucar.edu/knowledge_base/70549550), with 4 CPU Threads and 128gb of Memory

## Test 1
(40,000,000 Closed Polygons with 7 total verticies)

| Method / # of Polygons | 400,000 | 4,000,000 | 40,000,000 | 
| ------------- | ------------- | ------------- | ------------- |
| Construct with Exterior Coord Array  | 1.02s | 9.74s | 97.0s |
| Create Shapely Polygons + Construct with Shapely Polygons |  16.2s | 159s | 1560s
| Only Construct with Shapely Polygons |  15.2s | 149s | 1500s

Roughly a 16x speedup compared to using Shapely

### CODE REVIEW

**Comment by philipc2** [comment]
Depending on any initial feedback from the developers here, I'd be willing to continue implementing this to support Polygons with holes, MultiPolygons, etc.

While this PR focusses on Polygon's specifically, I believe other geometries could benefit from this type of construction support.

File: spatialpandas/geometry/polygon.py [hunk 1]
```diff
@@ -35,7 +37,6 @@ def construct_array_type(cls):
 
     @classmethod
     def _shapely_to_coordinates(cls, shape):
-        import shapely.geometry as sg
```
Comment (Hoxbro) [inline spatialpandas/geometry/polygon.py:38]: This is likely here to not make shapely a required dependencies.

Comment (philipc2) [reply spatialpandas/geometry/polygon.py:38]: Got it! Initially I presumed that this might lead to _some_ slowdown, but just testing it and it doesn't lead to any.

File: spatialpandas/geometry/multipolygon.py [hunk 2]
```diff
@@ -158,6 +159,36 @@ def from_geopandas(cls, ga, orient=True):
...
+
+        Note:
+            When using a Numpy Array is passed in, it is assumed that each entry is unfired, meaning that
+            all resulting MultiPolygons will have the same number of Polygons.
+
+        """
+        if isinstance(exterior_coords, (list, np.ndarray)):
+            if isinstance(exterior_coords, np.ndarray):
+                warnings.warn(f"Constructing a MultiPolygonArray using a Numpy Array with a fixed shape {exterior_coords.shape}."
+                              f"Each MultiPolygon will contain exactly {exterior_coords.shape[0]} Polygons")
```
Comment (Hoxbro) [inline spatialpandas/geometry/multipolygon.py:184]: This likely needs a stack level, usual `stacklevel=2`.

File: spatialpandas/geometry/multipolygon.py [hunk 3]
```diff
@@ -158,6 +159,36 @@ def from_geopandas(cls, ga, orient=True):
...
+            mpa = [[[arr.ravel()] for arr in exterior] for exterior in exterior_coords]
```
Comment (Hoxbro) [inline spatialpandas/geometry/multipolygon.py:185]: If a wrong input shape is passed as a `exterior_coords`, this could give a non-helpful error message.

**Review by Hoxbro — COMMENTED** [review]
The logic added seems simple enough not to add a lot of tests, but it would still be nice with one for both polygon and multi-polygon. 

Maybe also some tests that handle edge cases like wrong input shape/size.

**Review by Hoxbro — APPROVED** [review]

**Comment by Hoxbro** [comment]
Thank you for the PR! :+1:
